### DISCUSSION TITLE
Centralized orderbook + Fedimint custody

### DISCUSSION THREAD

**Discussion by justinmoon** [fedimint/fedimint#1784] [Fedimint Modules]
Lloyd Fournier (@LLFourn) pitched this idea a couple months ago https://www.youtube.com/watch?v=hCjbStBKCEQ&t=2278s

The idea is that the exchange could still benefit from the speed and scalability of a centralized orderbook, but also benefit from the federated custody of Fedimint. "Exchanges compete on orderbooks rather than on custody".

**Comment by LLFourn** [comment]
Yeah to expand on that you could *also* put the order book in the fedimint itself but this seems like creating a lot of consensus rounds for something that is not so security critical but is time critical.

In the custody only model the fedimint would custody the assets. When you want to make a trade you tell the fedimint "I want to exchange X BTC for Y USDT", the fedimint signs some message to this effect and escrows the funds. The order book verifies the signature. Then when the orderbook finds a match it goes to the fedimint and settles it immediately with the corresponding "I want to exchange Y USDT for X BTC" authorization from the matched user. The fedimint upon seeing these two matching orders (which it issued) can actually effect the transfer between accounts. 

The order book is still in charge of matching orders together. If the user wants to cancel an order they first go to the order book and get a signature that the order is cancelled. Then they take that and go to the fedimint to release their funds from the order. If the order book refuses to authorize the cancellation then you should be able to go to the fedimint directly and cancel it after some long timelock (which gives the order book time to effect any trades it had matched).

Hopefully that makes sense!

**Reply by justinmoon** [reply]
Would an orderbook like this be able to settle trades between separate fedimints? I'm assuming not ...

**Reply by LLFourn** [reply]
That's an interesting question. It would be straight forward to operate this as two separate order books for each fedimint but to just have arbitrageurs who were willing to hold balances on both fedimints operate on both order books. If the arbitrageur is the order book operator itself then perhaps it can even present a single unified order book.

**Reply by ecurrencyhodler** [reply]
Ethereum-based AMM's rely on abrigtrage for price discovery. If there's a premium, it will be found. This could also be a possible source of revenue for the fedimint hosting the orderbook.

**Comment by ecurrencyhodler** [comment]
What are your thoughts on pushing settlement to customer withdrawals from the exchange only? This means that the centralized orderbook + trades would happen outside of the fedimint and the only settling that would need to happen is if they want to withdraw? 

Settling trades within the fedimint at a high volume sounds really difficult to implement.

**Reply by LLFourn** [reply]
My guess is to do this without ever making the order book a custodian you'd have to employ some kind of state rollup ZKP system to "prove" your withdrawal was valid against a state commitment (or have the order book be able to monitor and prove invalid withdrawals are invalid). A lot of the hard work has been done to make this possible for ethereum based systems https://starkware.co/starkex/

I think it'd be better to wait for this to become an actual problem though. You can probably batch settlements together to some extent if there are users making lot back and forth trades between each other.

**Reply by justinmoon** [reply]
If settlement uses accounts then it might be much faster than if ecash is always used ...
